### ISSUE TITLE
Docker IPv6 (ip6tables) support breakes unrelated bridges

### ISSUE THREAD

**Issue by agners** [moby/moby#48365]
### Description

On a system running Docker 27.1.2 alongside with KVM/libvirt with it's own separate bridge, I've observed that IPv6 stopped working in virtual machines. It turns out that all IPv6 multicast packets don't make it through the bridge.

Ultimately, disabling filtering helped:
```
sysctl -w net.bridge.bridge-nf-call-ip6tables=0
```

I then was able to pinpoint Docker to be responsible. It seems that Docker adds a default `DROP` rule in the `FORWARDING` chain:

```
# ip6tables -L
...
Chain FORWARD (policy DROP)
...
```

When not starting Docker (or with `ip6tables` disabled), the default value of `ACCEPT` makes IPv6 in virtual machines just work fine.

I've also noticed that the IPv4 `FORWARD` chain uses default `ACCEPT`. 

Is the `DROP` for the IPv6 `FORWARD` chain intentional/required?

### Reproduce

1. Check `ip6tables -L` and observe the default for `FORWARD` rule is `ACCEPT`
2. Start docker with `--ip6tables true`
3. Check `ip6tables -L` and observe new default for `FORWARD` rule.

### docker version

```bash
Client:
 Version:           27.1.2
 API version:       1.46
 Go version:        go1.23.0
 Git commit:        d01f264bcc
 Built:             Sun Aug 18 14:43:14 2024
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          27.1.2
  API version:      1.46 (minimum version 1.24)
  Go version:       go1.23.0
  Git commit:       f9522e5e96
  Built:            Sun Aug 18 14:43:14 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.7.20
  GitCommit:        8fc6bcff51318944179630522a095cc9dbf9f353.m
 runc:
  Version:          1.1.13
  GitCommit:        
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Version:    27.1.2
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  0.16.2
    Path:     /usr/lib/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  2.29.2
    Path:     /usr/lib/docker/cli-plugins/docker-compose

Server:
 Containers: 135
  Running: 1
  Paused: 0
  Stopped: 134
 Images: 371
 Server Version: 27.1.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: journald
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 8fc6bcff51318944179630522a095cc9dbf9f353.m
 runc version: 
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.6.42-1-lts
 Operating System: Arch Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 32
 Total Memory: 62.01GiB
 Name: allenwind
 ID: 4YKC:JPTF:AI6R:AV34:KFP6:PR3Y:J72G:OEHE:3ZH2:OM3Y:6M52:HUFJ
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: agners
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by Mainfrezzer** [comment]
> I've also noticed that the IPv4 `FORWARD` chain uses default `ACCEPT`.
> 
> Is the `DROP` for the IPv6 `FORWARD` chain intentional/required?
> 


That behavior is indeed intentional since 27.0.3

Which i find funny, because this exact same situation happened with ipv4, which had some fallout, until a check for "net.ipv4.ip_forward=1" was implemented. One can hope that the same will happen for ipv6 at some point

**Comment by robmry** [comment]
> That behavior is indeed intentional since 27.0.3

The behaviour of `--ip6tables=true` hasn't changed but, since 27.0.1 it has been enabled by-default. (There was no 27.0.0, 27.0.1 was the major release.)

Running with `--ip6tables=false` will restore the old behaviour but then, of course, no `ip6tables` rules get set up.

> Which i find funny, because this exact same situation happened with ipv4, which had some fallout, until a check for "net.ipv4.ip_forward=1" was implemented. One can hope that the same will happen for ipv6 at some point

Hmm, right - you may mean these PRs ...
- https://github.com/moby/libnetwork/pull/2450
- https://github.com/moby/libnetwork/pull/2460

The current behaviour for IPv4, in [setupIPForwarding](https://github.com/moby/moby/blob/9fca5e7b6ff9aa29f85930d616565d5e1e09fad3/libnetwork/drivers/bridge/setup_ip_forwarding.go#L27-L74), is:
- if the daemon is running with `--ip-forward` (the default), and
- if the kernel's `net.ipv4.ip_forward = 0` when the daemon starts, then
- enable `net.ipv4.ip_forward` and set the `FORWARD` chain's default policy to `DROP`.

That seems wrong ... if something-else enables forwarding in the kernel before docker starts, container ports become accessible from remote hosts. The first of the PRs linked above attempted to fix that by setting the default policy to `DROP` whenever `--ip-forward=true` (regardless of whether docker itself set `net.ipv4.ip_forward = 1`). That sounds good, but the second PR reverts it... it doesn't link to any specific fallout, but it would have been a breaking change.

For IPv6 with `--ip6tables=true`, the default `DROP` policy has always been set when `--ip-forward=true`, regardless of initial kernel config.

It is weird that the default policy for the `FORWARD` chain is configured differently for IPv4/IPv6. But, I don't think we can just align them now, without breaking some set of use-cases:
- Making IPv4 behaviour more secure by re-applying the default policy fix above would cause the same breakage as last time.
- Removing the default `DROP` for IPv6 would make existing IPv6 networks less secure, including networks configured before 27.x with `--ip6tables=true` (as was recommended).

We could add options to override the default policies. But it'd still be confusing and, in practice, anything non-default is likely to get missed a lot of the time.

There's recently been some discussion between maintainers about changing the default behaviour to eliminate iptables rules for port-forwarding, relying on `docker-proxy` instead. (Because it's simpler, less intrusive in terms of host firewall policies, and may circumvent iptables vs. nftables issues a little. But, forwarding in user-space is less efficient, and `docker-proxy` is a man-in-the-middle so I don't think it copes with HTTPS/TLS.) Might that help? ... Probably not, we'd still need a `DROP` somewhere, or direct routing to container networks would still leave all ports exposed.

I'm wondering if we could just leave the default policy alone, and add per-bridge DROP rules to the DOCKER chain. So, for example, for a `-p 80` we'd add the current `ACCEPT` rule:
```
-A DOCKER -d 172.18.0.2/32 ! -i br-f297b9ef9be8 -o br-f297b9ef9be8 -p tcp -m tcp --dport 80 -j ACCEPT
```
And, always append a `DROP` for incoming traffic that hasn't been explicitly accepted:
```
-A DOCKER -d 172.18.0.2/32 ! -i br-f297b9ef9be8 -o br-f297b9ef9be8 -j DROP
```

Then...
- It'd be less intrusive, no change to the user's choice of default policy. Instead, `DOCKER` rules would only affect docker networks but, for those networks, I think it'd achieve the same as a default `DROP`.
- We could modify both IPv4 and IPv6 to have the same behaviour.
- It'd be a breaking change for IPv4 networks relying on direct-routing to a container from outside the host, with no port mappings set up. So, it'd be a major-release change, and maybe we'd want a way to disable the extra `DROP` rules so that people don't have to add explicit `-p`'s. But, I imagine there are a lot more networks where ports are unintentionally exposed to direct-routing because of the default `ACCEPT` for IPv4 - and doing something explicit to get less-secure behaviour is better than getting it by default.
- There'd be an extra rule per-bridge, which would have a marginal (but perhaps measurable) effect on max throughput. Probably acceptable if it sorts out the mess.

@akerouanton, @corhere, @cpuguy83, @thaJeztah - what am I missing, has this been suggested before?

**Comment by frebib** [comment]
> But, I imagine there are a lot more networks where ports are unintentionally exposed to direct-routing because of the default ACCEPT for IPv4

👋🏼 I'm definitely in the absolutely smallest minority here, but got here because I'm relying exactly on this behaviour, it seems. The recent IPv6 changes have been very exciting to me, but they've also been equally troublesome to make work. I'd love a flag/toggle for IPv4 & IPv6 to allow this behavior deliberately. In my case I don't use any port mappings at all, but rely on container<->outside the host communication _just working_ with gateway_mode_ipv6=routed. I suppose adding explicit port mappings would be the safer option, but having the ability to just open containers up to the wider network would also be nice, even if it's opt in. Bonus points if it can be made to work for the default bridge network too.
Perhaps this should be a separate issue; I'm not sure. I just landed here when I discovered that setting the default forward policy to ACCEPT made everything work again.

**Comment by robmry** [comment]
> > [...] networks relying on direct-routing to a container from outside the host, with no port mappings set up [...]. But, I imagine there are a lot more networks where ports are unintentionally exposed to direct-routing because of the default ACCEPT for IPv4
> 
> 👋🏼 I'm definitely in the absolutely smallest minority here, but got here because I'm relying exactly on this behaviour for both IPv4/6, it seems. The recent IPv6 changes have been very exciting to me, but they've also been equally troublesome to make work. I'd love a flag/toggle for IPv4 & IPv6 to allow this behavior deliberately. In my case I don't use any port mappings at all, but rely on container<->outside the host communication _just working_ with gateway_mode_ipv6=routed. I suppose adding explicit port mappings would be the safer option, but having the ability to just open containers up to the wider network would also be nice, even if it's opt in. Bonus points if it can be made to work for the default bridge network too. Perhaps this should be a separate issue; I'm not sure. I just landed here when I discovered that setting the default forward policy to ACCEPT made everything work again.

Hi @frebib - thank you for the use-case! This issue is the right place for it.

A `--gateway_mode_ipv[46]=routed-open` might do the trick ... it'd be like `routed` but wouldn't set up any per-port rules iptables rules, or the DROP described above. When a container's gateway network is in that mode, it'd probably reject port mapping options.

It wouldn't earn us the default bridge bonus points, as the default bridge doesn't yet have the `gateway_mode` options - but we may be able to sort that out separately as part of efforts to make the default bridge less unusual.

How does that sound?

**Comment by frebib** [comment]
Yeah that sounds like it would do the trick. Thanks

**Comment by jpiszcz** [comment]
I am using Docker with Debian Stable, what is the best way to add the following command for AFTER docker starts up?
`ip6tables -P FORWARD ACCEPT
`

**Comment by robmry** [comment]
> Yeah that sounds like it would do the trick. Thanks

Hi @frebib - just discussed this with moby network maintainers @corhere and @akerouanton ... we wondered if you'd considered using a docker "ipvlan" network to hook the container up to the network, instead of the bridge driver? The container would then get an interface that can talk directly to the local network, with no iptables to deal with. The ipvlan driver needs to know which host interface it should use, which may be a reason not to use it, but it would avoid routing via the docker host.

**Comment by Mainfrezzer** [comment]
I do want to mention that the default behavior of setting the default policy to drop, **despite being set by the user to accept before the docker start** is affecting other programs as well. Wireguard tunnels, LXC-Container etc.

Its quite frankly a pain in the ass that ive to setup a cron job to check if the default policy changed at all just because the docker service restarted or the host system.


This, this is the fallout i mentioned https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=865975, its literally the same issue just with the fancy new ipv6 paint.

Heres the pull https://github.com/moby/moby/pull/28257

**Comment by jpiszcz** [comment]
> I do want to mention that the default behavior of setting the default policy to drop, **despite being set by the user to accept before the docker start** is affecting other programs as well. Wireguard tunnels, LXC-Container etc.
> 
> Its quite frankly a pain in the ass that ive to setup a cron job to check if the default policy changed at all just because the docker service restarted or the host system.
> 
> This, this is the fallout i mentioned https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=865975, its literally the same issue just with the fancy new ipv6 paint.
> 
> Heres the pull #28257

Awesome idea, I am setting one up as well looking for drop:

```
# sudo nft list chain ip6 filter FORWARD
table ip6 filter {
        chain FORWARD {
                type filter hook forward priority filter; policy accept;
                counter packets 108520 bytes 195320256 jump DOCKER-USER
        }
}
```

**Comment by frebib** [comment]
> > Yeah that sounds like it would do the trick. Thanks
> 
> Hi @frebib - just discussed this with moby network maintainers @corhere and @akerouanton ... we wondered if you'd considered using a docker "ipvlan" network to hook the container up to the network, instead of the bridge driver? The container would then get an interface that can talk directly to the local network, with no iptables to deal with. The ipvlan driver needs to know which host interface it should use, which may be a reason not to use it, but it would avoid routing via the docker host.

I haven't really considered dabbling with other drivers to be honest. I can take a look. One immediate thought is that this trades off the firewall issue with packets not being able to be locally routed any more and always hairpinning via the next hop (unless I'm not understanding how this works.. I'll have to have a play?). I do think that there's probably merit to properly configuring port mappings with the bridge driver with `--gateway_mode_ipv[46]=routed`

**Comment by frebib** [comment]
I may be overstretching this issue now, but I've just discovered that container<->container communication is blocked by DOCKER-ISOLATION-STAGE-2 when using `--gateway_mode_ipv6=routed` even when ports are explicitly published for a given container. I'd have that down as a bug. My use-case is that I have containers with well-known addresses in DNS (humour me here) and some services communicate with others running in containers on the same host for various things; let's take Prometheus scrapes for example. My Prometheus server cannot scrape various containers on their well-known addresses because they don't happen to be in the same Docker network.. What's even more wild is that because of the routing, if these existed on separate hosts it *would* work. Now afaik this problem didn't exist in Docker v26.x (for IPv4 or IPv6, even) which I've been running for a while although I'm not quite sure what changed. Maybe Docker isn't intended to be used like this 🤔

**Comment by robmry** [comment]
> I may be overstretching this issue now, but I've just discovered that container<->container communication is blocked by DOCKER-ISOLATION-STAGE-2 when using `--gateway_mode_ipv6=routed` even when ports are explicitly published for a given container. I'd have that down as a bug. My use-case is that I have containers with well-known addresses in DNS (humour me here) and some services communicate with others running in containers on the same host for various things; let's take Prometheus scrapes for example. My Prometheus server cannot scrape various containers on their well-known addresses because they don't happen to be in the same Docker network.. What's even more wild is that because of the routing, if these existed on separate hosts it _would_ work. Now afaik this problem didn't exist in Docker v26.x (for IPv4 or IPv6, even) which I've been running for a while although I'm not quite sure what changed. Maybe Docker isn't intended to be used like this 🤔

Ah, yes - thank you @frebib ... with mode `routed` we had external access in mind, particularly for people who want to do IPv6 without any NAT. But it is weird that containers on neighbouring networks don't get access. It needs more thought - would you mind raising this one as a separate issue?

**Comment by robmry** [comment]
> I do want to mention that the default behavior of setting the default policy to drop, **despite being set by the user to accept before the docker start** is affecting other programs as well. Wireguard tunnels, LXC-Container etc.
> 
> Its quite frankly a pain in the ass that ive to setup a cron job to check if the default policy changed at all just because the docker service restarted or the host system.

Hi @Mainfrezzer - yes, that's what this issue is about. As you can see, we're working on it.

In the meantime, as noted above, setting `"ip6tables":false` in docker's `daemon.json` will restore the pre-27.x behaviour.

**Comment by jpiszcz** [comment]
Hello, I upgraded docker and now IPv6 is blocked again:
2024-12-09 16:06:00 configure docker-buildx-plugin:amd64 0.19.2-1~debian.12~bookworm <none>
2024-12-09 16:06:00 configure docker-compose-plugin:amd64 2.31.0-1~debian.12~bookworm <none>
2024-12-09 16:06:00 configure docker-ce-cli:amd64 5:27.4.0-1~debian.12~bookworm <none>
2024-12-09 16:06:00 configure docker-ce-rootless-extras:amd64 5:27.4.0-1~debian.12~bookworm <none>
2024-12-09 16:06:00 configure docker-ce:amd64 5:27.4.0-1~debian.12~bookworm <none>

`$ sudo nft list chain ip6 filter FORWARD
table ip6 filter {
        chain FORWARD {
                type filter hook forward priority filter; policy drop;
                counter packets 193 bytes 19492 jump DOCKER-USER
        }
}
`
